# Bootcamp Data Science y MLOps

<img src="https://i.ibb.co/5RM26Cw/LOGO-COLOR2.png" width="500px">

Creado en [escueladedatosvivos.ai](https://escueladedatosvivos.ai) 🚀.

¿Consultas? En la página tenés soporte por IA guiada, comunidad y el acceso a certificación.

<br>

---  

## 0) Dataset 👾

El dataset proviene de Kaggle: [Personal Key Indicators of Heart Disease](https://www.kaggle.com/datasets/kamilpytlak/personal-key-indicators-of-heart-disease).

## 1) Objetivo 

En base a los datos médicos se busca crear un modelo para predecir si una persona va a tener o no un ataque al corazón.

## 1) Cargamos los datos 📕

In [1]:
import pandas as pd

In [2]:
df_data = pd.read_csv("data/heart_2020_cleaned.csv", sep=',') 

In [3]:
df_data

HeartDisease    BMI Smoking AlcoholDrinking Stroke  PhysicalHealth  \
0                No  16.60     Yes              No     No             3.0   
1                No  20.34      No              No    Yes             0.0   
2                No  26.58     Yes              No     No            20.0   
3                No  24.21      No              No     No             0.0   
4                No  23.71      No              No     No            28.0   
...             ...    ...     ...             ...    ...             ...   
319790          Yes  27.41     Yes              No     No             7.0   
319791           No  29.84     Yes              No     No             0.0   
319792           No  24.24      No              No     No             0.0   
319793           No  32.81      No              No     No             0.0   
319794           No  46.56      No              No     No             0.0   

        MentalHealth DiffWalking     Sex  AgeCategory      Race Diabetic  \
0               30.0          No  Female        55-59     White      Yes   
1                0.0          No  Female  80 or older     White       No   
2               30.0          No    Male        65-69     White      Yes   
3                0.0          No  Female        75-79     White       No   
4                0.0         Yes  Female        40-44     White       No   
...              ...         ...     ...          ...       ...      ...   
319790           0.0         Yes    Male        60-64  Hispanic      Yes   
319791           0.0          No    Male        35-39  Hispanic       No   
319792           0.0          No  Female        45-49  Hispanic       No   
319793           0.0          No  Female        25-29  Hispanic       No   
319794           0.0          No  Female  80 or older  Hispanic       No   

       PhysicalActivity  GenHealth  SleepTime Asthma KidneyDisease SkinCancer  
0                   Yes  Very good        5.0    Yes            No        Yes  
1                   Yes  Very good        7.0     No            No         No  
2                   Yes       Fair        8.0    Yes            No         No  
3                    No       Good        6.0     No            No        Yes  
4                   Yes  Very good        8.0     No            No         No  
...                 ...        ...        ...    ...           ...        ...  
319790               No       Fair        6.0    Yes            No         No  
319791              Yes  Very good        5.0    Yes            No         No  
319792              Yes       Good        6.0     No            No         No  
319793               No       Good       12.0     No            No         No  
319794              Yes       Good        8.0     No            No         No  

[319795 rows x 18 columns]

*Nota:* si bien el valor defecto de sep en `read_csv` es la coma `,`. 
<br>Siempre lo hago explícito porque a veces los archivos vienen separados por punto y coma, u otro separador como tab. Es una buena práctica, y también aplica cuando graban archivos.

## 2) Inspección inicial 👀

In [4]:
df_data.columns

Index(['HeartDisease', 'BMI', 'Smoking', 'AlcoholDrinking', 'Stroke',
       'PhysicalHealth', 'MentalHealth', 'DiffWalking', 'Sex', 'AgeCategory',
       'Race', 'Diabetic', 'PhysicalActivity', 'GenHealth', 'SleepTime',
       'Asthma', 'KidneyDisease', 'SkinCancer'],
      dtype='object')

In [5]:
# Primeros 5 registros
df_data.head(5)

HeartDisease    BMI Smoking AlcoholDrinking Stroke  PhysicalHealth  \
0           No  16.60     Yes              No     No             3.0   
1           No  20.34      No              No    Yes             0.0   
2           No  26.58     Yes              No     No            20.0   
3           No  24.21      No              No     No             0.0   
4           No  23.71      No              No     No            28.0   

   MentalHealth DiffWalking     Sex  AgeCategory   Race Diabetic  \
0          30.0          No  Female        55-59  White      Yes   
1           0.0          No  Female  80 or older  White       No   
2          30.0          No    Male        65-69  White      Yes   
3           0.0          No  Female        75-79  White       No   
4           0.0         Yes  Female        40-44  White       No   

  PhysicalActivity  GenHealth  SleepTime Asthma KidneyDisease SkinCancer  
0              Yes  Very good        5.0    Yes            No        Yes  
1              Yes  Very good        7.0     No            No         No  
2              Yes       Fair        8.0    Yes            No         No  
3               No       Good        6.0     No            No        Yes  
4              Yes  Very good        8.0     No            No         No

In [6]:
# Los últimos 5 registros
df_data.tail(5)

HeartDisease    BMI Smoking AlcoholDrinking Stroke  PhysicalHealth  \
319790          Yes  27.41     Yes              No     No             7.0   
319791           No  29.84     Yes              No     No             0.0   
319792           No  24.24      No              No     No             0.0   
319793           No  32.81      No              No     No             0.0   
319794           No  46.56      No              No     No             0.0   

        MentalHealth DiffWalking     Sex  AgeCategory      Race Diabetic  \
319790           0.0         Yes    Male        60-64  Hispanic      Yes   
319791           0.0          No    Male        35-39  Hispanic       No   
319792           0.0          No  Female        45-49  Hispanic       No   
319793           0.0          No  Female        25-29  Hispanic       No   
319794           0.0          No  Female  80 or older  Hispanic       No   

       PhysicalActivity  GenHealth  SleepTime Asthma KidneyDisease SkinCancer  
319790               No       Fair        6.0    Yes            No         No  
319791              Yes  Very good        5.0    Yes            No         No  
319792              Yes       Good        6.0     No            No         No  
319793               No       Good       12.0     No            No         No  
319794              Yes       Good        8.0     No            No         No

In [7]:
df_data['Race'].value_counts()

White                             245212
Hispanic                           27446
Black                              22939
Other                              10928
Asian                               8068
American Indian/Alaskan Native      5202
Name: Race, dtype: int64

In [8]:
# https://pypi.org/project/funpymodeling/
from funpymodeling.exploratory import status

In [9]:
status(df_data)

variable  q_nan  p_nan  q_zeros   p_zeros  unique     type
0       HeartDisease      0    0.0        0  0.000000       2   object
1                BMI      0    0.0        0  0.000000    3604  float64
2            Smoking      0    0.0        0  0.000000       2   object
3    AlcoholDrinking      0    0.0        0  0.000000       2   object
4             Stroke      0    0.0        0  0.000000       2   object
5     PhysicalHealth      0    0.0   226589  0.708545      31  float64
6       MentalHealth      0    0.0   205401  0.642290      31  float64
7        DiffWalking      0    0.0        0  0.000000       2   object
8                Sex      0    0.0        0  0.000000       2   object
9        AgeCategory      0    0.0        0  0.000000      13   object
10              Race      0    0.0        0  0.000000       6   object
11          Diabetic      0    0.0        0  0.000000       4   object
12  PhysicalActivity      0    0.0        0  0.000000       2   object
13         GenHealth      0    0.0        0  0.000000       5   object
14         SleepTime      0    0.0        0  0.000000      24  float64
15            Asthma      0    0.0        0  0.000000       2   object
16     KidneyDisease      0    0.0        0  0.000000       2   object
17        SkinCancer      0    0.0        0  0.000000       2   object

**Veamos como se distribuye el variable target**

In [10]:
df_data['HeartDisease'].value_counts(dropna=False, normalize=True), df_data['HeartDisease'].value_counts(dropna=True, normalize=False)

(No     0.914405
 Yes    0.085595
 Name: HeartDisease, dtype: float64,
 No     292422
 Yes     27373
 Name: HeartDisease, dtype: int64)

Vemos que el dataset esta desbalanceado, teniendo que el 91.44% (292.422) de las personas no tuvieron un ataque al corazon y el %8.56 (27.373) de las personas si tuvieron un ataque al corazon.

Este desbalanceo puede desencadenar en un sesgo al `No`, por lo que hay que analizar bien que metrica escoger para evaluar el modelo.

## 3) Acondicionar dataset 🔧

Eliminamos las columnas `AgeCategory`, `GenHealth`, `Race`, `Diabetic`.
Esto lo hacemos con la finalidad de ver cómo influirán los demás parámetros en el análisis.

La columna `Race` es eliminada porque la consideramos un dato polemico y puede llegar a ser propenso a un sesgo/bias.

In [11]:
df_data_filter = df_data.drop(['AgeCategory', 'GenHealth', 'Race', 'Diabetic'], axis=1)

In [12]:
df_data_filter

HeartDisease    BMI Smoking AlcoholDrinking Stroke  PhysicalHealth  \
0                No  16.60     Yes              No     No             3.0   
1                No  20.34      No              No    Yes             0.0   
2                No  26.58     Yes              No     No            20.0   
3                No  24.21      No              No     No             0.0   
4                No  23.71      No              No     No            28.0   
...             ...    ...     ...             ...    ...             ...   
319790          Yes  27.41     Yes              No     No             7.0   
319791           No  29.84     Yes              No     No             0.0   
319792           No  24.24      No              No     No             0.0   
319793           No  32.81      No              No     No             0.0   
319794           No  46.56      No              No     No             0.0   

        MentalHealth DiffWalking     Sex PhysicalActivity  SleepTime Asthma  \
0               30.0          No  Female              Yes        5.0    Yes   
1                0.0          No  Female              Yes        7.0     No   
2               30.0          No    Male              Yes        8.0    Yes   
3                0.0          No  Female               No        6.0     No   
4                0.0         Yes  Female              Yes        8.0     No   
...              ...         ...     ...              ...        ...    ...   
319790           0.0         Yes    Male               No        6.0    Yes   
319791           0.0          No    Male              Yes        5.0    Yes   
319792           0.0          No  Female              Yes        6.0     No   
319793           0.0          No  Female               No       12.0     No   
319794           0.0          No  Female              Yes        8.0     No   

       KidneyDisease SkinCancer  
0                 No        Yes  
1                 No         No  
2                 No         No  
3                 No        Yes  
4                 No         No  
...              ...        ...  
319790            No         No  
319791            No         No  
319792            No         No  
319793            No         No  
319794            No         No  

[319795 rows x 14 columns]

**Valores categóricos a numéricos**

Las siguienes columnas:
- `Smoking`
- `AlcoholDrinking`
- `Stroke`
- `DiffWalking`
- `PhysicalActivity`
- `Asthma`
- `KidneyDisease`
- `SkinCancer`

Tiene solo dos posibles valores de tipo string (Yes/No). Por lo que vamos a covertirlas al tipo `int` (0/1).

In [13]:
class_map = {'No': 0, 'Yes': 1}
boolean_columns = ['Smoking', 'AlcoholDrinking', 'Stroke', 'DiffWalking', 'PhysicalActivity', 'Asthma', 'KidneyDisease', 'SkinCancer']

for boolean_column in boolean_columns:
    df_data_filter[boolean_column] = df_data_filter[boolean_column].map(class_map)

Para la columna `Sex` tenemos dos posibles valores de tipo string (Male/Female). Por lo que tambien vamos a covertirlas al tipo `int` (0/1).

In [14]:
class_gender_map = {'Male': 0, 'Female': 1}
df_data_filter['Sex'] = df_data_filter['Sex'].map(class_gender_map)

La columna `HeartDisease` es nuestro target.
Esta columna tiene dos posibles valores `Yes` (la persona tuvo un ataque al corazón) y `No` (la persona no tuvo un ataque al corazón). Por lo que tambien vamos a covertirlas al tipo `int` (0/1).

In [15]:
df_data_filter['HeartDisease']

0          No
1          No
2          No
3          No
4          No
         ... 
319790    Yes
319791     No
319792     No
319793     No
319794     No
Name: HeartDisease, Length: 319795, dtype: object

In [16]:
class_map = {'No': 0, 'Yes': 1}
df_data_filter['HeartDisease'] = df_data_filter['HeartDisease'].map(class_map)

In [17]:
df_data_filter.sample(5)

HeartDisease    BMI  Smoking  AlcoholDrinking  Stroke  PhysicalHealth  \
152835             0  30.45        1                0       0             0.0   
219299             0  30.54        0                0       0             0.0   
169024             0  19.20        0                0       0             0.0   
107784             0  33.12        0                0       0             0.0   
232404             0  28.48        0                0       0             0.0   

        MentalHealth  DiffWalking  Sex  PhysicalActivity  SleepTime  Asthma  \
152835          10.0            0    1                 1        8.0       0   
219299          10.0            0    0                 0        8.0       0   
169024           0.0            0    1                 1        7.0       0   
107784           5.0            0    0                 1       10.0       0   
232404           1.0            0    0                 1        7.0       0   

        KidneyDisease  SkinCancer  
152835              0           0  
219299              0           0  
169024              0           1  
107784              0           0  
232404              0           0

## 4) Separación input/output o X/Y ✂️

In [18]:
x_data = df_data_filter.drop('HeartDisease', axis=1)

In [19]:
x_data

BMI  Smoking  AlcoholDrinking  Stroke  PhysicalHealth  MentalHealth  \
0       16.60        1                0       0             3.0          30.0   
1       20.34        0                0       1             0.0           0.0   
2       26.58        1                0       0            20.0          30.0   
3       24.21        0                0       0             0.0           0.0   
4       23.71        0                0       0            28.0           0.0   
...       ...      ...              ...     ...             ...           ...   
319790  27.41        1                0       0             7.0           0.0   
319791  29.84        1                0       0             0.0           0.0   
319792  24.24        0                0       0             0.0           0.0   
319793  32.81        0                0       0             0.0           0.0   
319794  46.56        0                0       0             0.0           0.0   

        DiffWalking  Sex  PhysicalActivity  SleepTime  Asthma  KidneyDisease  \
0                 0    1                 1        5.0       1              0   
1                 0    1                 1        7.0       0              0   
2                 0    0                 1        8.0       1              0   
3                 0    1                 0        6.0       0              0   
4                 1    1                 1        8.0       0              0   
...             ...  ...               ...        ...     ...            ...   
319790            1    0                 0        6.0       1              0   
319791            0    0                 1        5.0       1              0   
319792            0    1                 1        6.0       0              0   
319793            0    1                 0       12.0       0              0   
319794            0    1                 1        8.0       0              0   

        SkinCancer  
0                1  
1                0  
2                0  
3                1  
4                0  
...            ...  
319790           0  
319791           0  
319792           0  
319793           0  
319794           0  

[319795 rows x 13 columns]

In [20]:
y_data = df_data_filter['HeartDisease']

In [21]:
y_data

0         0
1         0
2         0
3         0
4         0
         ..
319790    1
319791    0
319792    0
319793    0
319794    0
Name: HeartDisease, Length: 319795, dtype: int64

In [22]:
y_data.value_counts()

0    292422
1     27373
Name: HeartDisease, dtype: int64

## 5) Separación training / test 💪

In [23]:
from sklearn.model_selection import train_test_split 

Vamos a utilizar el 20% de los datos para el conjunto de test

In [24]:
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.2)

In [25]:
print(x_train.shape, y_train.shape)

(255836, 13) (255836,)


In [26]:
print(x_test.shape, y_test.shape)

(63959, 13) (63959,)


## 6) Creación del modelo: Random Forest 😎

In [27]:
# Importamos el modelo que queremos
from sklearn.ensemble import RandomForestClassifier 

In [28]:
# Creamos 100 decision trees (default)
classifier_model = RandomForestClassifier(n_estimators=100)

In [29]:
classifier_model

RandomForestClassifier()

In [30]:
%%time

classifier_model.fit(x_train, y_train)

CPU times: user 14 s, sys: 166 ms, total: 14.2 s
Wall time: 14.4 s


RandomForestClassifier()

## 7) Predecimos la clase ✅

In [31]:
# En training
pred_tr = classifier_model.predict(x_train)

In [32]:
# En testing
pred_ts = classifier_model.predict(x_test)

### 7.1 Validamos Training
Generamos un dataframe con la realidad vrs. lo predicho

In [33]:
df_val_tr = pd.DataFrame({'y_train': y_train, 'pred_tr': pred_tr})

In [34]:
df_val_tr

y_train  pred_tr
191642        0        0
34885         0        0
247079        0        0
247370        0        0
224730        0        0
...         ...      ...
250907        0        0
211385        0        0
306282        0        0
316958        0        0
297969        0        0

[255836 rows x 2 columns]

In [35]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [36]:
from pprint import pprint

In [37]:
# ¿cuántos aciertos?
train_metrics = {
    'accuracy': accuracy_score(df_val_tr.y_train, df_val_tr.pred_tr, normalize=True),
    'precision': precision_score(df_val_tr.y_train, df_val_tr.pred_tr),
    'recall': recall_score(df_val_tr.y_train, df_val_tr.pred_tr),
    'f1_score': f1_score(df_val_tr.y_train, df_val_tr.pred_tr),
}

pprint(train_metrics)

{'accuracy': 0.9788223705811535,
 'f1_score': 0.8615526140951602,
 'precision': 0.9836046443783184,
 'recall': 0.7664469197544896}


### 7.2 Validamos Testing
Generamos un dataframe con la realidad vrs. lo predicho

In [38]:
df_val_ts = pd.DataFrame({'y_test': y_test, 'pred_ts': pred_ts})

In [39]:
# ¿cuántos aciertos?
test_metrics = {
    'accuracy': accuracy_score(df_val_ts.y_test, df_val_ts.pred_ts, normalize=True),
    'precision': precision_score(df_val_ts.y_test, df_val_ts.pred_ts),
    'recall': recall_score(df_val_ts.y_test, df_val_ts.pred_ts),
    'f1_score': f1_score(df_val_ts.y_test, df_val_ts.pred_ts),
}

pprint(test_metrics)

{'accuracy': 0.8973717537797652,
 'f1_score': 0.1575975359342916,
 'precision': 0.2543496271748136,
 'recall': 0.11416883599851246}


## Extra 🎯

### Variar número de estimadores

In [40]:
# Creamos 5 decision trees
classifier_model_1 = RandomForestClassifier(n_estimators=5)

In [41]:
classifier_model_1

RandomForestClassifier(n_estimators=5)

In [42]:
%%time

classifier_model_1.fit(x_train, y_train)

CPU times: user 736 ms, sys: 37.4 ms, total: 774 ms
Wall time: 795 ms


RandomForestClassifier(n_estimators=5)

In [43]:
# En testing
pred_ts_1 = classifier_model_1.predict(x_test)

In [44]:
df_val_ts_1 = pd.DataFrame({'y_test': y_test, 'pred_ts': pred_ts_1})

In [45]:
df_val_ts_1

y_test  pred_ts
3506         0        0
28715        0        0
30786        0        0
23836        0        0
287731       0        0
...        ...      ...
291531       0        0
161911       0        0
103636       0        0
284739       0        0
280967       0        0

[63959 rows x 2 columns]

In [46]:
# ¿cuántos aciertos?
test_metrics = {
    'accuracy': accuracy_score(df_val_ts_1.y_test, df_val_ts_1.pred_ts, normalize=True),
    'precision': precision_score(df_val_ts_1.y_test, df_val_ts_1.pred_ts),
    'recall': recall_score(df_val_ts_1.y_test, df_val_ts_1.pred_ts),
    'f1_score': f1_score(df_val_ts_1.y_test, df_val_ts_1.pred_ts),
}

pprint(test_metrics)

{'accuracy': 0.8887881298957144,
 'f1_score': 0.167388505208943,
 'precision': 0.2259083728278041,
 'recall': 0.13294905169207885}


### Variar número de procesadores

In [47]:
# Creamos 100 decision trees (default)
classifier_model_2 = RandomForestClassifier(n_estimators=5, n_jobs=-1)

In [48]:
classifier_model_2

RandomForestClassifier(n_estimators=5, n_jobs=-1)

In [49]:
%%time

classifier_model_2.fit(x_train, y_train)

CPU times: user 982 ms, sys: 58.6 ms, total: 1.04 s
Wall time: 303 ms


RandomForestClassifier(n_estimators=5, n_jobs=-1)

In [50]:
# En testing
pred_ts_2 = classifier_model_2.predict(x_test)

In [51]:
df_val_ts_2 = pd.DataFrame({'y_test': y_test, 'pred_ts': pred_ts_2})

In [52]:
df_val_ts_2

y_test  pred_ts
3506         0        0
28715        0        0
30786        0        0
23836        0        0
287731       0        0
...        ...      ...
291531       0        0
161911       0        0
103636       0        0
284739       0        0
280967       0        0

[63959 rows x 2 columns]

In [53]:
# ¿cuántos aciertos?
test_metrics = {
    'accuracy': accuracy_score(df_val_ts_2.y_test, df_val_ts_2.pred_ts, normalize=True),
    'precision': precision_score(df_val_ts_2.y_test, df_val_ts_2.pred_ts),
    'recall': recall_score(df_val_ts_2.y_test, df_val_ts_2.pred_ts),
    'f1_score': f1_score(df_val_ts_2.y_test, df_val_ts_2.pred_ts),
}

pprint(test_metrics)

{'accuracy': 0.8912584624525086,
 'f1_score': 0.1723194097346186,
 'precision': 0.2393388429752066,
 'recall': 0.13462253625883228}


Podemos apreciar que al utilizar todos los procesadores disponibles dentro de la computadora para un número de estimadores de 500.
El tiempo de entrenamiento se reduce de 13.9 s (13900ms) a 929 ms, lo cual es una reduccion de un poco mas del 93%.

Pero en termino de accuracy no hay mucho cambio.

Tomamos varias metricas del modelo `accuracy`, `precision`, `recall`, `f1_score`.

Esto con el find de tener mas nocion del comportamiento del modelo.

### Conclusiones

Con la metrica puede parecer que el modelo funciona correctamente, pero la realidad es que no (mirando el resto de metricas.)

Esto puede deberse al desbalanceo del dataset en funcion del `target`.

**Tengan cuidado** cuando tengan dataset con miles de filas, puede llegar a tomar mucho tiempo en visualizar del dataframe o colapsar el notebook. Así como nos paso en la clase del martes 😥😁.
<br>Para eso usen esta configuración de Pandas:

In [ ]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

Veamos cómo se ve un dataframe completo ahora

In [ ]:
status(df_data)